File: otros_scripts/prof_analysis.R
library(RMySQL)

calendar_commit_month_author <- function(year, month, author){
	M <- rep(0, 31)
	if (length(author) == 0) return(M)
	for (i in 1:length(author[,1])){
		if ((as.numeric(format(strptime(author[i,2], "%Y-%m-%d %H:%M:%S"),'%m')) == month) &
			(as.numeric(format(strptime(author[i,2], "%Y-%m-%d %H:%M:%S"),'%Y')) == year)){
			M[as.numeric(format(strptime(author[i,2], "%Y-%m-%d %H:%M:%S"),'%d'))] <- M[as.numeric(format(strptime(author[i,2], "%Y-%m-%d %H:%M:%S"),'%d'))] + 1
		}
	}
	return(M)
}

con <- dbConnect(MySQL(),
		user="tthebosss", password="1234",
		dbname="openstack_authors", host="localhost")


rs <- dbSendQuery(con, 'SELECT MIN(date) FROM scmlog')
dat_min <- as.numeric(format(strptime(fetch(rs, n=-1)[1,1], "%Y-%m-%d %H:%M:%S"),'%Y'))
dbClearResult(rs)

rs <- dbSendQuery(con, 'SELECT MAX(date) FROM scmlog')
dat_max <- as.numeric(format(strptime(fetch(rs, n=-1)[1,1], "%Y-%m-%d %H:%M:%S"),'%Y'))
dbClearResult(rs)

rs <- dbSendQuery(con, 'SELECT * FROM companies')
companies <- fetch(rs, n=-1)
dbClearResult(rs)

rs <- dbSendQuery(con, 'SELECT upeople_id, company_id FROM upeople_companies WHERE upeople_id = ANY (SELECT upeople_id FROM people_upeople)')
upeople_companies <- fetch(rs, n=-1)
dbClearResult(rs)

rs <- dbSendQuery(con, 'SELECT people_id FROM people_upeople WHERE upeople_id = ANY 
						(SELECT upeople_id FROM upeople_companies)')
people <- fetch(rs, n=-1)
dbClearResult(rs)
tot_people <- length(people[,1])

authors <- list()
for (people in 1:tot_people){
	query <- sprintf('SELECT author_id, date FROM scmlog WHERE author_id = %s ORDER BY date', people)
	rs <- dbSendQuery(con, query)
	authors[people] <- list(fetch(rs, n=-1))
}

work_authors <- list()
for (author in 1:tot_people){
	print(c('Author', author))
	M <- c()
	for (year in dat_min:dat_max){
		for (month in 1:12){
			aut <- calendar_commit_month_author(year, month, authors[author][[1]])
			work_days <- 0
			for (i in 1:length(aut)){
				if (aut[i] != 0){
					work_days <- work_days + 1
				}
			}
			M[length(M)+1] <- work_days
		}
	#print(c('Vector rellenado', M))
	work_authors[author] <- M
	}
}

save(work_authors, 'work_authors.RData')
print(work_authors)

"
print (people[,1])
print (colnames(people))

print(dim(upeople_companies)[1])
print(tot_people)
print(fetch(dbSendQuery(con, 'SELECT count(*) from people')))
"

dbClearResult(rs)
dbDisconnect(con)

File: otros_scripts/prof_analysis.py
#!/usr/bin/python
# -*- coding: utf-8 -*- 

"""
Description: Analisis de los autores de openstack para detectar patrones de clasificacion en profesionales
"""

import MySQLdb
from datetime import *
import numpy as np

def calendar_commit_month_author(year, month, author):
    M = np.zeros(31, dtype=np.int)
    for aux in author:
        if (int(aux[2].month) == month) & (int(aux[2].year) == year):
            M[int(aux[2].day-1)] += 1
    return M

con = MySQLdb.connect(host='localhost', user='tthebosss', passwd='1234', \
                      db='openstack_authors')
cursor = con.cursor()

cursor.execute('SELECT MIN(date) FROM scmlog')
date_min = cursor.fetchall()[0][0]
date_min = date_min.year
cursor.execute('SELECT MAX(date) FROM scmlog')
date_max = cursor.fetchall()[0][0]
date_max = date_max.year

period = range(date_min,date_max)
period.append(date_max)

cursor.execute('SELECT * FROM companies')
list_companies = cursor.fetchall()
tot_companies = len(list_companies)

companies = np.zeros([tot_companies, 2], dtype='a16')
for comp in range(0, tot_companies):
    companies[comp][0] = list_companies[comp][1]

cursor.execute('SELECT company_id FROM upeople_companies WHERE upeople_id = ANY (SELECT upeople_id FROM people_upeople)')
upeople_company = cursor.fetchall()

for people in upeople_company:
    if companies[people[0]-1][1] != '':
        companies[people[0]-1][1] = str(int(companies[people[0]-1][1]) + 1)
    else:
        companies[people[0]-1][1] = 1

query = ('SELECT people_id FROM people_upeople WHERE upeople_id = ANY '
         '(SELECT upeople_id FROM upeople_companies)')
cursor.execute(query)
authors_id = cursor.fetchall()
tot_authors = len(authors_id)

authors = []
for i in range(0,tot_authors):
    query = ('SELECT committer_id, author_id, date FROM scmlog '
              'WHERE author_id = %s ORDER BY date')
    cursor.execute(query, authors_id[i][0])
    authors.append(cursor.fetchall())

print len(authors)
work_authors = []
a = 0
for author in authors:
    M = []
    a += 1
    print a
    file = open('info_openstack/' + str(a), 'w')
    for year in period:
        for month in range(1, 13):
            aut = calendar_commit_month_author(year, month, author)
            file.write(str(aut))
            file.write('\n')
            work_days = 0
            for i in range(0, len(aut)):
                if aut[i] != 0:
                    work_days += 1
            M.append(work_days)
    work_authors.append(M)
    file.close()

aux = 0
aux_sum = 0
for i in range(0, len(work_authors)):
    j = np.sum(work_authors[i])
    if j > aux_sum:
        aux = i
        aux_sum = j

print work_authors[aux]

cursor.close()
con.close()